File: twoPointers/01-100-twoSum.py

class Solution:
    def twoSumHashTable(self,nums,target):
        d=dict()
        for i,v in enumerate(nums):
            if target-v in d:
                return [d[target-v],i]
            d[v]=i
            
    def twoSum(self, nums, target):
        v1,v2=0,0
        bkNums=nums[::]
        nums.sort()
        n=len(nums)
        l,r=0,n-1
        while l<r:
            if nums[l]+nums[r]==target:
                v1=nums[l]
                v2=nums[r]
                break
            elif nums[l]+nums[r]<target:
                l+=1
            else:
                r-=1
        ans=[]
        for i in range(n):
            if bkNums[i] in (v1,v2):
                ans.append(i)
                if len(ans)==2:
                    return ans
nums=[-1,-2,-3,-4,-5]
# bkNums=nums[::]
# nums.sort()

print(Solution().twoSum(nums,-8))
nums=[-1,-2,-3,-4,-5]

print(Solution().twoSumHashTable(nums,-8))

File: map/01-100-twoSum.cpp
#include<map>
#include<iostream>
#include<vector>

using VEC=std::vector<int>;
using std::cout;

using MAP=std::map<int,int>;


VEC twoSum(const VEC& nums,int target)
{
    MAP d{};
    VEC ans{};
    int n=nums.size();
    for(int i=0;i<n;++i)
    {
        int cur=nums[i];
        if(d.find(target-cur)!=d.end())
        {
            ans.push_back(d[target-cur]);
            ans.push_back(i);
            return ans;
        }
        d[cur]=i;
    }
    return ans;
}

int main()
{
    VEC nums{-1,-2,-3,-4,-5};
    VEC res=twoSum(nums,-8);
    for(auto &v:res)
        cout<<v<<',';
}